### PR TITLE
MarcioHuser's Alpakit changes

### PR DESCRIPTION
**Pull request by budak7273**
@MarcioHuser's Alpakit changes from #76

- Added checkbox to select multiple Plugins
- Added 'Alpakit All!' button to deploys all checked Plugins in one go

(cherry picked from commit 0e68d580d2ccaa73a7171ceeafdc0d0abac30b12)

### CODE REVIEW

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp [hunk 1]
```diff
@@ -1,9 +1,51 @@
...
+	return GetMutableDefault<UAlpakitSettings>();
 }
 
 void UAlpakitSettings::SaveSettings() {
-    SaveConfig();
+	SaveConfig();
+}
+
+
+UAlpakitModSelectionSettings* UAlpakitModSelectionSettings::Get() {
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp:14]: Can you please explain what's the point in making completely new settings just for your additions? Why can't you just reuse normal alpakit settings?

File: Plugins/Alpakit/Source/Alpakit/Public/AlpakitModEntry.h [hunk 2]
```diff
@@ -11,9 +11,16 @@ class SAlpakitModEntry : public SCompoundWidget {
 
 	void Construct(const FArguments& Args, TSharedRef<IPlugin> InMod, TSharedRef<SAlpakitModEntryList> InOwner);
 	
-	void PackageMod() const;
+	void PackageMod(const TArray<TSharedPtr<SAlpakitModEntry>>& nextEntries) const;
+	void OnEnableCheckboxChanged(ECheckBoxState NewState);
+
+	inline bool IsSelected()
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Public/AlpakitModEntry.h:17]: Use FORCEINLINE if you write implementations in headers

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp [hunk 3]
```diff
@@ -1,9 +1,51 @@
...
+	return GetMutableDefault<UAlpakitModSelectionSettings>();
+}
+
+void UAlpakitModSelectionSettings::SaveSettings() {
+	SaveConfig();
+}
+
+void UAlpakitModSelectionSettings::loadJson() {
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp:22]: Just... why? Why did you need all of that stuff? Do you know TMap<FString, bool> can be marked as properties too and saved/persisted? What's your motivation behind that code?

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp [hunk 4]
```diff
@@ -108,3 +120,48 @@ void SAlpakitModEntryList::SetShowEngine(bool bInShowEngine) {
...
 	LoadMods();
 }
+
+FReply SAlpakitModEntryList::PackageAllMods() {
+	TSharedPtr<SAlpakitModEntry> first;
+	TArray<TSharedPtr<SAlpakitModEntry>> nextEntries;
+
+	UE_LOG(LogTemp, Display, TEXT("Alpakit All!"))
+
+	for (auto mod : FilteredMods) {
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:130]: do not use auto, use proper type. Also, here you are implicitly copying mod object, because you're using auto instead of proper reference type. 
Also, codestyle. Variable names start with uppercase.

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp [hunk 5]
```diff
@@ -108,3 +120,48 @@ void SAlpakitModEntryList::SetShowEngine(bool bInShowEngine) {
...
+
+		auto modEntry = StaticCastSharedPtr<SAlpakitModEntry>(tableRow->GetContent());
+		if (!modEntry.IsValid()) {
+			UE_LOG(LogTemp, Display, TEXT("TableRow content is not valid!"))
+
+			continue;
+		}
+
+		if(!modEntry->IsSelected()) {
+			UE_LOG(LogTemp, Display, TEXT("Plugin is not selected: %s"), * mod->GetName());
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:148]: Is that an abnormal condition? do you really need to log it?

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp [hunk 6]
```diff
@@ -80,14 +92,14 @@ bool PluginMatchesSearchTokens(const IPlugin& Plugin, const TArray<FString>& Tok
 void SAlpakitModEntryList::Filter(const FString& InFilter) {
 	LastFilter = InFilter;
 	FilteredMods.Empty();
-	
-	if (InFilter.IsEmpty()) {
+
+	if (InFilter.IsEmpty()) 	{
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:96]: these random space -> tab changes. fix them.

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp [hunk 7]
```diff
@@ -60,34 +81,60 @@ FText GetCurrentPlatformName() {
...
+									return x != nextMod;
+								})
+							);
+					});
+			});
+}
+
+void SAlpakitModEntry::OnEnableCheckboxChanged(ECheckBoxState NewState) {
+	// Save new checked state for the mod at the
+	auto modSelectionSettings = UAlpakitModSelectionSettings::Get();
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp:134]: Auto again

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp [hunk 8]
```diff
@@ -60,34 +81,60 @@ FText GetCurrentPlatformName() {
...
 
 	const FText PlatformName = GetCurrentPlatformName();
 	IUATHelperModule::Get().CreateUatTask(CommandLine, PlatformName,
-			LOCTEXT("PackageModTaskName", "Packaging Mod"),
-        LOCTEXT("PackageModTaskShortName", "Package Mod Task"),
-        FAlpakitStyle::Get().GetBrush("Alpakit.OpenPluginWindow"));
+		LOCTEXT("PackageModTaskName", "Packaging Mod"),
+		LOCTEXT("PackageModTaskShortName", "Package Mod Task"),
+		FAlpakitStyle::Get().GetBrush("Alpakit.OpenPluginWindow"),
+		nextEntries.Num() == 0
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp:114]: Why not move it all to the intermediate variable?

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp [hunk 9]
```diff
@@ -1,35 +1,56 @@
 #include "AlpakitModEntry.h"
-
-
 #include "AlpakitSettings.h"
 #include "AlpakitStyle.h"
+#include "IPlatformFilePak.h"
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp:4]: do you even need that include?

**Review by Archengius — CHANGES_REQUESTED** [review]

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp (hunk 3 above)
Comment (MarcioHuser) [reply Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp:22]: Using tmap caused UE crash 🤷🏽‍♂️

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp (hunk 5 above)
Comment (MarcioHuser) [reply Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:148]: It was made for debugging purposely. Left overs

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp (hunk 9 above)
Comment (MarcioHuser) [reply Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp:4]: Maybe a leftover auto insurance include from rider

File: Plugins/Alpakit/Source/Alpakit/Public/AlpakitModEntry.h [hunk 10]
```diff
@@ -11,9 +11,16 @@ class SAlpakitModEntry : public SCompoundWidget {
 
 	void Construct(const FArguments& Args, TSharedRef<IPlugin> InMod, TSharedRef<SAlpakitModEntryList> InOwner);
 	
-	void PackageMod() const;
+	void PackageMod(const TArray<TSharedPtr<SAlpakitModEntry>>& NextEntries) const;
+	void OnEnableCheckboxChanged(ECheckBoxState NewState);
+
+	FORCEINLINE bool IsSelected()
+	{
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Public/AlpakitModEntry.h:18]: Codestyle. First bracket should be on the same line as method declaration.

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp [hunk 11]
```diff
@@ -1,9 +1,11 @@
...
+#include "Json.h"
+
 UAlpakitSettings* UAlpakitSettings::Get() {
-    return GetMutableDefault<UAlpakitSettings>();
+	return GetMutableDefault<UAlpakitSettings>();
 }
 
 void UAlpakitSettings::SaveSettings() {
-    SaveConfig();
+	SaveConfig();
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp:10]: Use spaces, not tabs

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitWidget.cpp [hunk 12]
```diff
@@ -31,28 +31,28 @@ void SAlpakitWidget::Construct(const FArguments& InArgs) {
...
-        		.Content()[
-        			SNew(STextBlock)
-        			.Text(LOCTEXT("CheckAllPlugins", "All Content Plugins"))
-        		]
-        		.OnCheckStateChanged_Lambda([this](ECheckBoxState InState) {
-        			this->ModList->SetShowEngine(InState == ECheckBoxState::Checked);
-        		})
-        	]
-        ]
+			DetailsView.ToSharedRef()
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitWidget.cpp:34]: Use spaces, not tabs

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp [hunk 13]
```diff
@@ -108,3 +120,48 @@ void SAlpakitModEntryList::SetShowEngine(bool bInShowEngine) {
...
+
+	UE_LOG(LogTemp, Display, TEXT("Alpakit All!"))
+
+	for (TSharedRef<IPlugin> Mod : FilteredMods) {
+		UE_LOG(LogTemp, Display, TEXT("Collecting Plugin %s!"), *Mod->GetName())
+
+		TSharedPtr<ITableRow> TableRow = ModList->WidgetFromItem(Mod);
+		if (!TableRow.IsValid()) {
+			UE_LOG(LogTemp, Display, TEXT("TableRow not found!"))
+
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:136]: Unnecessary newlines

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp [hunk 14]
```diff
@@ -108,3 +120,48 @@ void SAlpakitModEntryList::SetShowEngine(bool bInShowEngine) {
...
+			continue;
+		}
+
+		TSharedPtr<SAlpakitModEntry> ModEntry = StaticCastSharedPtr<SAlpakitModEntry>(TableRow->GetContent());
+		if (!ModEntry.IsValid()) {
+			UE_LOG(LogTemp, Display, TEXT("TableRow content is not valid!"))
+
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:143]: Unnecessary newlines

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp [hunk 15]
```diff
@@ -108,3 +120,48 @@ void SAlpakitModEntryList::SetShowEngine(bool bInShowEngine) {
...
+			continue;
+		}
+
+		if(!ModEntry->IsSelected()) {
+			UE_LOG(LogTemp, Display, TEXT("Plugin is not selected: %s"), * Mod->GetName());
+
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:149]: Unnecessary newlines

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp [hunk 16]
```diff
@@ -60,34 +80,62 @@ FText GetCurrentPlatformName() {
...
 
 	const FText PlatformName = GetCurrentPlatformName();
 	IUATHelperModule::Get().CreateUatTask(CommandLine, PlatformName,
-			LOCTEXT("PackageModTaskName", "Packaging Mod"),
-        LOCTEXT("PackageModTaskShortName", "Package Mod Task"),
-        FAlpakitStyle::Get().GetBrush("Alpakit.OpenPluginWindow"));
+		LOCTEXT("PackageModTaskName", "Packaging Mod"),
+		LOCTEXT("PackageModTaskShortName", "Package Mod Task"),
+		FAlpakitStyle::Get().GetBrush("Alpakit.OpenPluginWindow"),
+		NextEntries.Num() == 0
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp:113]: Please rewrite this, nesting levels here are NOT okay

**Review by Archengius — CHANGES_REQUESTED** [review]
Formatting is still lacking

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp [hunk 17]
```diff
@@ -1,93 +1,129 @@
...
+        AdditionalUATArguments.Append(TEXT("-CopyToGameDir "));
+    }
+    if (Settings->LaunchGameAfterPacking != EAlpakitStartGameType::NONE && NextEntries.Num() == 0) {
+        AdditionalUATArguments.Append(TEXT("-LaunchGame "));
+        AdditionalUATArguments.Append(GetArgumentForLaunchType(Settings->LaunchGameAfterPacking)).Append(TEXT(" "));
+    }
+
+    const FString LaunchGameArgument = GetArgumentForLaunchType(Settings->LaunchGameAfterPacking);
+
+    UE_LOG(LogTemp, Display, TEXT("Packaging plugin \"%s\". %d remaining"), *PluginName, NextEntries.Num());
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntry.cpp:93]: Use LogAlpakit, not LogTemp. LogTemp is not intended to be used outside of local debugging ever.

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp [hunk 18]
```diff
@@ -1,5 +1,7 @@
 #include "AlpakitSettings.h"
 
+#include "Json.h"
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitSettings.cpp:3]: Unneeded extra include

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp [hunk 19]
```diff
@@ -1,110 +1,164 @@
...
 }
+
+FReply SAlpakitModEntryList::PackageAllMods() {
+    TSharedPtr<SAlpakitModEntry> First;
+    TArray<TSharedPtr<SAlpakitModEntry>> NextEntries;
+
+    UE_LOG(LogTemp, Display, TEXT("Alpakit All!"));
+
+    for (TSharedRef<IPlugin> Mod : FilteredMods) {
+        UE_LOG(LogTemp, Display, TEXT("Collecting Plugin %s!"), *Mod->GetName());
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:131]: Use LogAlpakit and not LogTemp

File: Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp [hunk 20]
```diff
@@ -1,110 +1,164 @@
...
-	LoadMods();
+    bShowEngine = bInShowEngine;
+    LoadMods();
 }
+
+FReply SAlpakitModEntryList::PackageAllMods() {
+    TSharedPtr<SAlpakitModEntry> First;
+    TArray<TSharedPtr<SAlpakitModEntry>> NextEntries;
+
+    UE_LOG(LogTemp, Display, TEXT("Alpakit All!"));
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Private/AlpakitModEntryList.cpp:128]: LogAlpakit

File: Plugins/Alpakit/Source/Alpakit/Alpakit.Build.cs [hunk 21]
```diff
@@ -21,7 +21,8 @@ public Alpakit(ReadOnlyTargetRules Target) : base(Target)
 				"CoreUObject",
 				"Engine",
 				"Slate",
-				"SlateCore"
+				"SlateCore",
+				"Json",
```
Comment (Archengius) [inline Plugins/Alpakit/Source/Alpakit/Alpakit.Build.cs:25]: Unnecessary module dependency

**Review by Archengius — CHANGES_REQUESTED** [review]
